### PR TITLE
stats: automatically delete stats for dropped tables

### LINKED ISSUE #94195: Statistics for deleted tables in system.table_statistics do not get removed
**Describe the problem**

The problem was first noticed when performance of `SHOW TABLES` on a specific cluster slowed to > 60 seconds. A statement trace showed that most time was spent on the following query to the `system.table_statistics` table:

```
SELECT
  s."tableID", max(s."rowCount")
FROM
  system.table_statistics AS s
  JOIN (
      SELECT "tableID", max("createdAt") AS last_dt FROM system.table_statistics GROUP BY "tableID"
    )
      AS l ON l."tableID" = s."tableID" AND l.last_dt = s."createdAt"
    AS OF SYSTEM TIME '-10s'
GROUP BY
  s."tableID"
```

The `system.table_statistics` table had grown quite large (> 5 million rows) and appeared to retain statistics on deleted tables.

**To Reproduce**

Start a single node.

```
cockroach start-single-node --insecure
```

Query for the count of rows in `system.table_statistics` by database name. A watch provides repeated output of the command.

```
./cockroach sql --execute 'select t.database_name, count(ts."tableID") from system.table_statistics ts left join crdb_internal.tables t on ts."tableID" = t.table_id group by t.database_name;' --watch 10s --insecure
```
(it may take a couple of minutes for the statistics to show for system tables)

Set the `gc.ttlseconds` on the default range to `500` seconds to force garbage collection of the table:

```
ALTER RANGE DEFAULT CONFIGURE ZONE USING gc.ttlseconds = 300;
```

Create a database and table:

```
CREATE DATABASE d;
CREATE TABLE d.t (k uuid, v string);
```

Observe that rows in `system.table_statistics` are created for the table.
(it may take a couple of minutes for the statistics to show for the newly created table)

Drop the database:

```
SET sql_safe_updates=false;
DROP DATABASE d;
```

Observe that the rows for the table are not deleted, even after waiting until garbage collection completes (5 minutes).

**Expected behavior**
When a table is dropped, either directly or via dropping a database, the related rows in `system.table_statistics` are deleted. This prevents accumulation of orphaned rows that can slow down the performance of `SHOW TABLES`.

**Additional data / screenshots**
None

**Environment:**
 - CockroachDB version: reproduced in 21.2.12 and 22.1.1
 - Server OS: reproduced on MacOS but also observed on Linux
 - Client app: cockroach sql

**Additional context**
May impact other queries or cluster operations.

Jira issue: CRDB-22733

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit adds another task to the stats refresher that periodically runs a query to delete stats for dropped tables from the system table. By default, this query runs once an hour, but this can be configured via a cluster setting (which also exposes a way to disable this new "stats garbage collector" when it is set to 0). The query also limits the number of dropped tables to process at once to 1000 by default (controled via another cluster setting). The rationale for introducing the limit is to prevent a huge DELETE when the cluster that has been running for long time with many dropped tables has just upgraded to the binary with this fix.

Fixes: #94195.

Release note (bug fix): CockroachDB now automatically deletes statistics for dropped tables from `system.table_statistics` table.

### CODE REVIEW

**Review by rytaft — COMMENTED** [review]
Thanks for doing this!

> we currently don't have a way to easily force schema GC on the dropped tables

Is there no way to make the GC interval smaller as a testing knob? It would be nice to have some sort of non-manual test so we don't break this feature in the future...

> Fixes: [#94915](https://github.com/cockroachdb/cockroach/pull/94915).

This doesn't look like the right issue

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-:-NYZdyFA3V_ef_gusnFW:b-xemea4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/stats/automatic_stats.go` line 601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-NYZdGQc3RzYiChHbfWV:-NYZdGQc3RzYiChHbfWW:blqubye) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2767383743265126159cd3772cfe8b956ee9c75/pkg/sql/stats/automatic_stats.go#L601)):*
> ```Go
> 				select {
> 				case <-intervalChangedCh:
> 					continue
> ```

Is it worth adding some logging here (and for the same case below)?

**Review by yuzefovich — COMMENTED** [review]
> Is there no way to make the GC interval smaller as a testing knob? It would be nice to have some sort of non-manual test so we don't break this feature in the future...

I asked about this on [slack](https://cockroachlabs.slack.com/archives/C0168LW5THS/p1686614721060869), and my conclusion from that thread is that it is currently not easy to speed up GC of the dropped table. To go around it, I've just introduced a testing knob and added the corresponding test. It'd be good to get sign off from @chengxiong-ruan on this.

> This doesn't look like the right issue

Indeed, nice catch, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-:-NYZpqx_4F0REpCSjXKi:b-kax5jl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @michae2, and @rytaft)

___
*[`pkg/sql/stats/automatic_stats.go` line 601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-NYZdGQc3RzYiChHbfWV:-NYZrRa95i6MDtm9foaK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2767383743265126159cd3772cfe8b956ee9c75/pkg/sql/stats/automatic_stats.go#L601)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is it worth adding some logging here (and for the same case below)?
</blockquote></details>

Done.

File: pkg/sql/schema_changer.go [hunk 1]
```diff
@@ -2326,6 +2326,10 @@ type GCJobTestingKnobs struct {
 
 	// If true, the GC job will not wait for MVCC GC.
 	SkipWaitingForMVCCGC bool
+
+	// If true, disables the usage of DelRange (i.e. falls back to the legacy
+	// way).
+	DisableDelRange bool
```
Comment (chengxiong-ruan) [inline pkg/sql/schema_changer.go:2332]: Could add a TODO (to me) to remove this when we figure out how to control the mvcc gc queue? Thanks.

**Review by chengxiong-ruan — APPROVED** [review]
gc jobs change lgtm.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105364)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @michae2, and @rytaft)

___
*[`pkg/sql/schema_changer.go` line 2332 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105364#gh-1239073319:-NY_88JM9u0xsL8q2fQH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25cc72bde5583b83b74307c53c6357e90e344b3f/pkg/sql/schema_changer.go#L2332)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Could add a TODO (to me) to remove this when we figure out how to control the mvcc gc queue? Thanks.
</blockquote></details>

Done.

**Comment by erikgrinaker** [comment]
The `DisableDelRange` testing knob can be replaced with this:

```diff
diff --git a/pkg/sql/stats/delete_stats_dropped_table_test.go b/pkg/sql/stats/delete_stats_dropped_table_test.go
index 7208606f8a5..ae723b572bb 100644
--- a/pkg/sql/stats/delete_stats_dropped_table_test.go
+++ b/pkg/sql/stats/delete_stats_dropped_table_test.go
@@ -31,11 +31,7 @@ func TestStatsAreDeletedForDroppedTables(t *testing.T) {
        defer log.Scope(t).Close(t)
 
        params, _ := tests.CreateTestServerParams()
-       params.Knobs.GCJob = &sql.GCJobTestingKnobs{
-               // We disable the usage of DeleteRange primitive in order to speed up
-               // the test.
-               DisableDelRange: true,
-       }
+       params.ScanMaxIdleTime = time.Millisecond // speed up MVCC GC queue scans
        s, sqlDB, _ := serverutils.StartServer(t, params)
        defer s.Stopper().Stop(context.Background())
        runner := sqlutils.MakeSQLRunner(sqlDB)
@@ -44,6 +40,10 @@ func TestStatsAreDeletedForDroppedTables(t *testing.T) {
        runner.Exec(t, "SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;")
        // Lower the garbage collection interval to speed up the test.
        runner.Exec(t, "SET CLUSTER SETTING sql.stats.garbage_collection_interval = '1s';")
+       // Poll for MVCC GC more frequently.
+       runner.Exec(t, "SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '1s';")
+       // Cached protected timestamp state delays MVCC GC, update it every second.
+       runner.Exec(t, "SET CLUSTER SETTING kv.protectedts.poll_interval = '1s';")
 
        // Create a table with short TTL and collect stats on it.
        runner.Exec(t, "CREATE TABLE t (k PRIMARY KEY) AS SELECT 1;")
```

**Review by yuzefovich — COMMENTED** [review]
Thanks Erik for figuring this out! Updated the PR with the suggestion.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-:-NYchp_ZDxaZqrdUt9y3:b-a8g6i1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @michae2, and @rytaft)

**Review by rytaft — APPROVED** [review]
:lgtm: thanks for adding the test!

> Reviewed 2 of 5 files at r2, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-:-NYeXNhK5Nlpj97_RazR:b-dory68)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @michae2, and @yuzefovich)

___
*[`pkg/sql/stats/automatic_stats.go` line 601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-NYZdGQc3RzYiChHbfWV:-NYeWsUt6ytVzmM0mwXm:buvo2o4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2767383743265126159cd3772cfe8b956ee9c75/pkg/sql/stats/automatic_stats.go#L601)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

How about for the interval changed? Or do we already log changes to cluster settings?

**Review by yuzefovich — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-:-NYehvanDNHg0-K_HbJr:b-tlx7ko)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @michae2, and @rytaft)

___
*[`pkg/sql/stats/automatic_stats.go` line 601 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105364#-NYZdGQc3RzYiChHbfWV:-NYehx4dDnFRrdGuuLQZ:be6j5rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2767383743265126159cd3772cfe8b956ee9c75/pkg/sql/stats/automatic_stats.go#L601)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

How about for the interval changed? Or do we already log changes to cluster settings?
</blockquote></details>

We don't, but I don't think we log setting changes in other similar situations (e.g. statement diagnostics registry has "poling interval", but we don't explicitly log when the interval is changed via the cluster setting), so I'll keep it out here too.

**Pull request merged**